### PR TITLE
docker: move default from 24.x to 27.x, docker_24 mark as vulnerable, docker_{25,26}: bump

### LINKED ISSUE #328275: Why is the default docker version still at 24?
## Issue description

Today one of my containers wouldn't start after I updated the docker-compose file and it was due to the docker engine being too old to support a certain feature.

Easy enough to fix in config but is there a specific reason the `docker` attribute points at `docker_24` rather than `docker_27`?

Last time the attr was touched was https://github.com/NixOS/nixpkgs/commit/2318de70529d5b15293037943e470c1d77c4cb2c

@offlinehacker @vdemeester @periklis @teutat3s @otavio @guoard

### PR DESCRIPTION
**Pull request by teutat3s**
## Description of changes

`24.x` is [no longer maintained](https://github.com/moby/moby/pull/46772#discussion_r1686464084) as of February 1, 2024.
It did not (yet?) receive a fix for [CVE-2024-41110](https://github.com/moby/moby/security/advisories/GHSA-v23v-6jw2-98fq) .

Fixes #328275.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by kirillrdy — APPROVED** [review]
`docker.tests` pass :-)

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -29916,7 +29916,7 @@ with pkgs;
   inherit (callPackage ../applications/virtualization/docker {})
     docker_24 docker_25 docker_26 docker_27;
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:29917]: We should at least mark docker_24 as insecure.

Comment (teutat3s) [reply pkgs/top-level/all-packages.nix:29917]: Good point. I'm back at a computer on Monday. Anybody with availability feel free to take over if you like.

**Comment by SuperSandro2000** [comment]
let me push the meta.knownVulnerabilities for all of those and another round of bumps

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by Atemu** [comment]
This must be backported somehow and that's a bit tricky since 24.05 released with 24 as the default.

We must either:

1. Patch v24 to fix the CVE
2. Mark Docker as vulnerable on 24.05
3. Bump Docker to a version >v24 on 24.05

cc @wegank @Mic92

**Comment by Atemu** [comment]
Ah, just saw the backport PR above.

**Pull request merged**